azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java

```diff
@@ -142,6 +149,61 @@ private MySQLBasicDataSource(String host, int port, String dbName,
       setTestOnBorrow(true);
     }
 
+    /**
+     * Get a singleton object for MySQL BasicDataSource
+     */
+    public static MySQLBasicDataSource getInstance(String host, int port, String dbName,
+        String user, String password, int numConnections) {
+      if (instance == null) {
+        synchronized (MySQLBasicDataSource.class) {
+          if (instance == null) {
+            logger.info("Instantiating MetricReportManager");
+            instance = new MySQLBasicDataSource(host, port, dbName, user, password, numConnections);
+          }
+        }
+      }
+      return instance;
+    }
+
+    /**
+     * This method overrides {@link BasicDataSource#getConnection()}, in order to have retry logics.
+     *
+     */
+    @Override
+    public synchronized Connection getConnection() throws SQLException {
+
+      /*
+       * when az server process launches, if the connection isn't valid (e.g., network problem, wrong password, etc),
+       * we fail the connection immediately.
+       */
+      if (getDataSource() == null) {
+        return createDataSource().getConnection();
+      }
+
+      Connection connection = null;
+      int retryAttempt = 0;
+      while (connection == null && retryAttempt < ServerInternals.MAX_DB_RETRY_COUNT) {
+        try {
+          /*
+           * when DB connection could not be fetched here, dbcp library will keep searching until a timeout defined in
+           * its code hardly.
+           */
+          connection = createDataSource().getConnection();
+          if(connection != null)
+            return connection;
+        } catch (SQLException ex) {
+          this.dataSource = null;
+          logger.error("Failed to find DB connection. waits 1 minutes and retry. No.Attempt = " + retryAttempt, ex);
+        } catch (Exception ex) {
+          logger.error("Exception other than SQLException is thrown", ex);
+          return null;
+        } finally {
+          retryAttempt ++;
```

Comment:
Will putting this statement inside the catch (SQLException ex) { block have the same effect?